### PR TITLE
feat: add ranged monster

### PR DESCRIPTION
**Pull request by JoenvandeVorle**
## Description

Added a ranged monster prefab that shoots at the player from a distance. The monster has a "visibility range" and an "attack range", and also checks for a line of sight to the player before shooting.

## Testability

Open the `RangedEnemy` scene to play around with a respawning enemy. Once the player is within visibility, the enemy should start coming closer. Once the player is within attack range and the enemy has a direct line of sight, the enemy will attack. If the enemy loses line of sight but is still within range, it will start coming closer to the player. If the player leaves the visibility range, the enemy will stop altogether.
There is also a `RangedEnemyTestScene` accompanied by an aplib test.

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch
- [x] Added a scene to the game for your changes

### CODE REVIEW

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs [hunk 1]
```diff
@@ -22,8 +24,7 @@ protected override void Awake()
     {
         base.Awake();
         _meleeWeapon = GetComponentInChildren<MeleeWeapon>();
-        _meleeWeapon.TargetTag = _targetTag;
-        _meleeWeapon.Damage = _damagePoints;
+        _meleeWeapon.Initialize(_damagePoints, _targetTag, _swingLength, _swingWidth);
         _cooldownTimer = new(_attackCooldown);
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs:28]: This changes after updating with main

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs [hunk 2]
```diff
@@ -39,6 +34,20 @@ private void Awake()
...
+    /// <summary>
+    /// Initialize the weapon with the damage, height, radius and target tag.
+    /// </summary>
+    /// <param name="damage">The amount of damage the weapon will deal.</param>
+    /// <param name="height">The height of the hitzone in world units.</param>
+    /// <param name="radius">The radius of the hitzone in world units.</param>
+    /// <param name="targetTag">The tag of the target.</param>
+    public void Initialize(int damage, string targetTag, float height, float radius)
+    {
+        base.Initialize(damage, targetTag);
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs:46]: Not sure but it says base is not necessary here, also it is not an override

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs [hunk 3]
```diff
@@ -9,6 +9,8 @@ public class MeleeEnemy : DummyEnemy
 {
     [SerializeField] private float _attackCooldown = 2f;
     [SerializeField] private float _hitDelay = 1f;
+    [SerializeField] private float _swingWidth = 1.4f;
+    [SerializeField] private float _swingLength = 4f;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs:13]:
```suggestion
    [SerializeField]
    private float _attackCooldown = 2f;
    
    [SerializeField]
    private float _hitDelay = 1f;
    
    [SerializeField]
    private float _swingWidth = 1.4f;
    
    [SerializeField]
    private float _swingLength = 4f;
    
```

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs [hunk 4]
```diff
@@ -9,8 +9,35 @@
...
+    /// <summary>
+    /// Initialize the weapon with the fire point, damage, range and target tag.
+    /// </summary>
+    /// <param name="firePoint">The point from which the weapon will be fired.</param>
+    /// <param name="damage">The amount of damage the weapon will deal.</param>
+    /// <param name="range">The range of the weapon.</param>
+    /// <param name="targetTag">The tag of the target.</param>
+    public void Initialize(int damage, string targetTag, Transform firePoint, int range)
+    {
+        base.Initialize(damage, targetTag);
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs:37]: Same base comment here.

**Review by QuakeEye — CHANGES_REQUESTED** [review]

**Review by QuakeEye — COMMENTED** [review]
I'm not able to review much yet as a lot of things need changing when updating with main

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
+using UnityEngine;
+using System.Collections;
+
+public class RangedEnemy : DummyEnemy
+{
+    [SerializeField] private int _attackRange = 10;
+    [SerializeField] private int _visionRange = 20;
+    [SerializeField] private float _attackCooldown = 2f;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs:8]:
```suggestion
    [SerializeField]
    private int _attackRange = 10;
    
    [SerializeField]
    private int _visionRange = 20;
    
    [SerializeField]
    private float _attackCooldown = 2f;
    
```

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs [hunk 6]
```diff
@@ -9,8 +9,35 @@
 /// </summary>
 public class RangedWeapon : Weapon
 {
-    [SerializeField] private int _damage = 50;
+    /// <summary>
+    /// The point from which the weapon will be fired.
+    /// </summary>
+    [SerializeField] private Transform _firePoint;
     [SerializeField] private int _range = 50;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs:16]:
```suggestion
    [SerializeField] private Transform _firePoint;
    
    [SerializeField] private int _range = 50;
    
```

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/Weapon.cs [hunk 7]
```diff
@@ -8,7 +8,21 @@ public abstract class Weapon : MonoBehaviour
     /// <summary>
     /// The tag of the target that the weapon can hit.
     /// </summary>
-    public string TargetTag = "Enemy";
+    [SerializeField] protected string _targetTag = "Enemy";
+
+    /// <summary>
+    /// The amount of damage the weapon deals.
+    /// </summary>
+    [SerializeField] protected int _damage = 25;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/Weapon.cs:16]:
```suggestion
    [SerializeField]
    protected string _targetTag = "Enemy";

    /// <summary>
    /// The amount of damage the weapon deals.
    /// </summary>
    [SerializeField]
    protected int _damage = 25;
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
I got summoned by @QuakeEye to join in and be annoying about attribute formatting (tip: enable Resharper, then it's done automatically)

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs (hunk 2 above)
Comment (JoenvandeVorle) [reply aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs:46]: I put base here to make clearer that it is the method from the abstract weapon class, but I can remove that if not necessary. It is not an override because it's a different method with a different access modifier.

Comment (JensSteenmetz) [reply aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs]: I agree with @JoenvandeVorle. Clarity is love, clarity is life.

File: aplib.net-demo/Assets/Prefabs/Player/Player.prefab
Comment (QuakeEye) [inline aplib.net-demo/Assets/Prefabs/Player/Player.prefab:1]: Changing this prefab has reapplied the melee weapon to the 73_... scene, this is the main scene, please remove the melee weapon from that scene under the player

File: aplib.net-demo/Assets/Prefabs/Enemies/Melee Enemy.prefab
Comment (QuakeEye) [inline aplib.net-demo/Assets/Prefabs/Enemies/Melee Enemy.prefab:1]: The melee enemy appears to be broken now, please check out the main scene 73_...

File: aplib.net-demo/Assets/Prefabs/Player/Player.prefab
Comment (QuakeEye) [reply aplib.net-demo/Assets/Prefabs/Player/Player.prefab:1]: This is actually also the case in your own scene

File: aplib.net-demo/Assets/Prefabs/Enemies/Target Dummy.prefab [hunk 8]
```diff
@@ -312,7 +312,6 @@ MonoBehaviour:
   m_Name: 
   m_EditorClassIdentifier: 
   _maxHealth: 100
-  _entitySound: {fileID: 2844816295009711930}
 --- !u!114 &4628526191064518394
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Prefabs/Enemies/Target Dummy.prefab:315]: Why is entitysound removed from here?

File: aplib.net-demo/Assets/Scenes/RangedEnemy.unity
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scenes/RangedEnemy.unity:1]: Yeah so make sure to fix the double weapons problem here as well, as well as in the main scene 73_...

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs [hunk 9]
```diff
@@ -2,32 +2,45 @@
...
+    private float _swingWidth = 1.4f;
+
+    [SerializeField]
+    private float _swingLength = 4f;
+
     private readonly float _sizeIncrease = 1.2f;
     private MeleeWeapon _meleeWeapon;
     private Timer _cooldownTimer;
     private bool _isSwinging = false;
+    private EntitySound _entitySound;
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs:29]: We do not want to store the EntitySound and use it in this class, requiring it to enforce it existing on an enemy is fine, but it should be used within the weapons as they were, so that they can be used more easily and do not have to be used for every enemy type we may introduce

(Right now it's easier, melee weapon = melee sound, ranged weapon = ranged sound, check this out on the main branch!)

this comment applies to all uses of entity sound in this script

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs [hunk 10]
```diff
@@ -71,4 +84,13 @@ private IEnumerator SwingCoroutine()
...
     }
+
+    /// <summary>
+    /// Play the death sound when the enemy dies.
+    /// </summary>
+    protected override void OnDeath(HealthComponent _healthComponent)
+    {
+        base.OnDeath(_healthComponent);
+        _entitySound.OnDeath();
+    }
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs:95]: Yes so this function should thus not be necessary in this class

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    [SerializeField]
+    private int _attackRange = 10;
+
+    [SerializeField]
+    private int _visionRange = 20;
+
+    private bool _movingCloser;
+    private Timer _attackTimer;
+    private RangedWeapon _rangedWeapon;
+    private EntitySound _entitySound;
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs:25]: We do not want to store the EntitySound and use it in this class, requiring it to enforce it existing on an enemy is fine, but it should be used within the weapons as they were, so that they can be used more easily and do not have to be used for every enemy type we may introduce

(Right now it's easier, melee weapon = melee sound, ranged weapon = ranged sound, check this out on the main branch!)

this comment applies to all uses of entity sound in this script

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            _attackTimer.Reset();
+        }
+
+        base.Update();
+    }
+
+    /// <summary>
+    /// Coroutine to move closer until the enemy is within line of sight.
+    /// </summary>
+    /// <returns></returns>
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs:82]: This line can probably be removed if it is not filled out, otherwise people will get documentation like `Returns: `

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    /// <summary>
+    /// Play the death sound when the enemy dies.
+    /// </summary>
+    protected override void OnDeath(HealthComponent healthComponent)
+    {
+        base.OnDeath(healthComponent);
+        _entitySound.OnDeath();
+    }
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs:102]: Yeah so again, this should not be necessary here as it is handled in the weapons themselves

File: aplib.net-demo/Assets/Scripts/Entities/Player/PlayerLogic.cs [hunk 14]
```diff
@@ -6,11 +6,13 @@
 [RequireComponent(typeof(HealthComponent))]
 [RequireComponent(typeof(Movement))]
 [RequireComponent(typeof(ResetRigidbody))]
+[RequireComponent(typeof(EntitySound))]
 public class PlayerLogic : MonoBehaviour
 {
     private HealthComponent _healthComponent;
     private ResetRigidbody _resetRigidbody;
-    
+    private EntitySound _entitySound;
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Player/PlayerLogic.cs:14]: We do not want to store the EntitySound and use it in this class, requiring it to enforce it existing on an enemy is fine, but it should be used within the weapons as they were, so that they can be used more easily and do not have to be used for every enemy type we may introduce

(Right now it's easier, melee weapon = melee sound, ranged weapon = ranged sound, check this out on the main branch!)

this comment applies to all uses of entity sound in this script

File: aplib.net-demo/Assets/Scripts/Entities/Player/PlayerLogic.cs [hunk 15]
```diff
@@ -36,6 +39,7 @@ private void OnDeath(HealthComponent healthComponent)
         Debug.Log("Player died!");
         _resetRigidbody.ResetObject();
         _healthComponent.Reset();
+        _entitySound.OnDeath();
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Player/PlayerLogic.cs:42]: This should not be here, it is handled in the HealthComponent

File: aplib.net-demo/Assets/Scripts/Entities/Player/PlayerLogic.cs (hunk 14 above)
Comment (QuakeEye) [reply aplib.net-demo/Assets/Scripts/Entities/Player/PlayerLogic.cs:14]: (Copy paste but this applies to weapon+health component as well)

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs [hunk 16]
```diff
@@ -5,14 +5,45 @@
 namespace Entities.Weapons
 {
     /// <summary>
-    /// This class has functions for ranged weapons.
+    /// This class has fuctions for ranged weapons.
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs:8]: Why did you bring back the spelling mistake? Please make sure to merge things properly or ask for help if you are having issues, there have been multiple things seemingly brought back that were fixed on main after you created your branch :/

File: aplib.net-demo/Assets/Scripts/Entities/EntitySound.cs [hunk 17]
```diff
@@ -18,17 +17,12 @@ public class EntitySound : MonoBehaviour
...
     /// <summary>
     /// Sets up the AudioSource component.
     /// </summary>
     private void Awake()
     {
         _audioSource = GetComponent<AudioSource>();
-        _healthComponent = GetComponent<HealthComponent>();
-
-        _healthComponent.Death += OnDeath;
     }
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/EntitySound.cs:26]: Why did you remove this? Subscribing a Death sound to the health component was the better way to do this, this seems like a wrong merge again?

File: aplib.net-demo/Assets/Scripts/Entities/PathFind.cs [hunk 18]
```diff
@@ -36,5 +36,33 @@ public void UpdateAgent()
...
     }
+
+    /// <summary>
+    /// Set the distance at which the agent will stop from the goal.
+    /// </summary>
+    /// <param name="distance"></param>
+    public void SetStoppingDistance(float distance)
+    {
+        _agent.stoppingDistance = distance;
+    }
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/PathFind.cs:47]: Can be a lambda function

File: aplib.net-demo/Assets/Scripts/Entities/PathFind.cs [hunk 19]
```diff
@@ -36,5 +36,33 @@ public void UpdateAgent()
...
+
+    /// <summary>
+    /// Start/stop the agent.
+    /// </summary>
+    /// <param name="enabled">True to enable the agent, false to disable it.</param>
+    public void ToggleAgent(bool enabled)
+    {
+        _agent.isStopped = !enabled;
+    }
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/PathFind.cs:56]: Can be a lambda function

File: aplib.net-demo/Assets/Scripts/Entities/PathFind.cs [hunk 20]
```diff
@@ -36,5 +36,33 @@ public void UpdateAgent()
...
+
+    /// <summary>
+    /// Check if the goal is within a certain range.
+    /// </summary>
+    /// <param name="range">The range to check.</param>
+    /// <returns>True if the goal is within the range, false otherwise.</returns>
+    public bool GoalWithinRange(float range)
+    {
+        return Vector3.Distance(transform.position, _goal.position) <= range;
+    }
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/PathFind.cs:66]: Can be a lambda function

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+using UnityEngine;
+using UnityEngine.SceneManagement;
+using UnityEngine.TestTools;
+
+namespace Testing.AplibTests
+{
+    public class RangedEnemyBeliefSet : BeliefSet
+    {
+        public static readonly int FramesToRemember = 300;
+        /// <summary>
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:20]:
```suggestion
        public static readonly int FramesToRemember = 300;

        /// <summary>
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+            // Goal: Reach the target position 
+            PrimitiveGoalStructure<RangedEnemyBeliefSet> moveGoal = new(goal: new Goal<RangedEnemyBeliefSet>(moveTactic, MovePredicate));
+
+            // Goal: Wait until killed by the enemy
+            PrimitiveGoalStructure<RangedEnemyBeliefSet> waitGoal = new(goal: new Goal<RangedEnemyBeliefSet>(waitTactic, PlayerRespawnedPredicate));
+
+            // Goal: Wait and check if enemy is not following player
+            PrimitiveGoalStructure<RangedEnemyBeliefSet> waitGoal2 = new(goal: new Goal<RangedEnemyBeliefSet>(waitTactic, EnemyNotMovingPredicate));
+
+            // PrimitiveGoalStructure<RangedEnemyBeliefSet> attackGoal = new(goal: new Goal<RangedEnemyBeliefSet>(attackTactic, EnemyKilledPredicate));
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:95]: What is this? Does it need to be removed?

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+using Aplib.Integrations.Unity.Actions;
+using NUnit.Framework;
+using System.Collections;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+using UnityEngine.TestTools;
+
+namespace Testing.AplibTests
+{
+    public class RangedEnemyBeliefSet : BeliefSet
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:17]: There are quite a couple of comments that need a period at the end to satisfy our rules

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+
+                return Vector3.Distance(player.transform.position, target) < _targetStoppingDistance;
+            }
+
+            bool PlayerRespawnedPredicate(RangedEnemyBeliefSet beliefSet)
+            {
+                GameObject player = beliefSet.Player;
+                Vector3 respawnPosition = beliefSet.RespawnPosition;
+
+                return Vector3.Distance(player.transform.position, respawnPosition) < 0.5f;
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:129]: Should this number be defined at the start of this class?

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+            }
+
+            bool EnemyNotMovingPredicate(RangedEnemyBeliefSet beliefSet)
+            {
+                Vector3 target = beliefSet.TargetPosition;
+                Vector3 targetMemory = beliefSet.TargetPosition.GetMemoryAt(RangedEnemyBeliefSet.FramesToRemember - 1, true);
+                return Vector3.Distance(target, targetMemory) < 0.05f;
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:136]: Should this number be defined at the start of this class?

**Review by QuakeEye — CHANGES_REQUESTED** [review]

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs (hunk 9 above)
Comment (JoenvandeVorle) [reply aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs:29]: Yes I agree for the weapons but this is for the OnDeath soundeffects. In hindsight, I disagree with the way death was done in the sound PR. I think it is cleaner for the death soundeffects to be handled in the enemies themselves then for EntitySound to require a HealthComponent and listen to the death event. This way you can still have enemies with sound (from the weapons) but without a death sound .

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs (hunk 24 above)
Comment (JoenvandeVorle) [reply aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:129]: I do not think that's necessary in this case. It might even be clearer to have it defined there.

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs (hunk 25 above)
Comment (JoenvandeVorle) [reply aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:136]: Same as previous.

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs (hunk 9 above)
Comment (JensSteenmetz) [reply aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs:29]: I somewhat agree with your point @JoenvandeVorle; the entities shouldn't always implement a death sound. I don't think it should be done in the enemies themselves either, because in my opinion that violates the modular architecture that Unity supports. Basically, the EntitySound script should make this more flexible (allowing people to link sound clips to events through the inspector, ideally). The EntitySound script is a temporary fix, however, as sound clips will be mostly handled by the animation system later. I'm not sure if the EntitySound script is at all necessary at that point. For now, until the sound is done through the animation system, I think both of your solutions are valid.

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs (hunk 24 above)
Comment (JensSteenmetz) [reply aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs]: You could maybe give a name to it for extra clarity, but it's decently clear from the method name what you are checking imo.

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs [hunk 26]
```diff
@@ -13,19 +13,22 @@ public class MeleeWeapon : Weapon
         /// <summary>
         /// The amount of damage the weapon deals.
         /// </summary>
-        [SerializeField] public int Damage = 25;
+        [SerializeField]
+        public int Damage = 25;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs:17]: Is it correct that we have a Damage field and a _damage field now? Either way, that should probably change. Also, quick tip for later: you can make a public auto-property serialized in the inspector like so:
```suggestion
        [field: SerializeField]
        public int Damage { get; set; } = 25;
```

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs [hunk 27]
```diff
@@ -11,8 +11,39 @@ namespace Entities.Weapons
...
+        private int _range = 50;
+
+        private IEnumerable<RaycastHit> _orderedHits;
+
+        /// <summary>
+        /// By default, assume the weapon will be fired by the player, from the camera.
+        /// </summary>
+        private void Awake()
+        {
+            if (_firePoint == null)
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs:30]: Might want to do the actual null-check here? But I might be wrong.
```suggestion
            if (_firePoint is null)
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 28]
```diff
@@ -0,0 +1,139 @@
...
+using System.Collections;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+using UnityEngine.TestTools;
+
+namespace Testing.AplibTests
+{
+    public class RangedEnemyBeliefSet : BeliefSet
+    {
+        public static readonly int FramesToRemember = 300;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:19]: Seems like a constant?
```suggestion
        public const int FramesToRemember = 300;
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            Debug.Log($"Starting {nameof(RangedEnemyAplibTest)}");
+            SceneManager.LoadScene(_sceneName);
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            Debug.Log($"Finished {nameof(RangedEnemyAplibTest)}");
+            SceneManager.UnloadSceneAsync(_sceneName);
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:61]: Shouldn't use this. Unity does this automatically.
```suggestion
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        [UnityTest]
+        public IEnumerator PerformRangedEnemyTest()
+        {
+            InputManager.Instance.enabled = false;
+            RangedEnemyBeliefSet beliefSet = new();
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:68]: To prevent conflicts with my other suggestion.
```suggestion
            RangedEnemyBeliefSet rangedEnemyBeliefSet = new();
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 31]
```diff
@@ -0,0 +1,139 @@
...
+
+            // Create an intent for the agent that moves the agent towards the target position.
+            TransformPathfinderAction<RangedEnemyBeliefSet> transformPathfinderAction = new(
+                b =>
+                {
+                    GameObject player = b.Player;
+                    return player.GetComponent<Rigidbody>();
+                },
+                beliefSet.TargetPosition,
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:77]:
```suggestion
                beliefSet => beliefSet.TargetPosition,
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 32]
```diff
@@ -0,0 +1,139 @@
...
+                0.9f
+            );
+
+            // Tactic: Wait and do nothing.
+            PrimitiveTactic<RangedEnemyBeliefSet> waitTactic = new(new Action<RangedEnemyBeliefSet>(b => { }));
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:82]:
```suggestion
            PrimitiveTactic<RangedEnemyBeliefSet> waitTactic = new(new Action<RangedEnemyBeliefSet>(_ => { }));
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 33]
```diff
@@ -0,0 +1,139 @@
...
+
+            // Tactic: Move the player towards the target position.
+            PrimitiveTactic<RangedEnemyBeliefSet> moveTactic = new(transformPathfinderAction);
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:85]:
```suggestion
            PrimitiveTactic<RangedEnemyBeliefSet> moveTowardsTarget = new(transformPathfinderAction);
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 34]
```diff
@@ -0,0 +1,139 @@
...
+
+            // Goal: Reach the target position.
+            PrimitiveGoalStructure<RangedEnemyBeliefSet> moveGoal = new(goal: new Goal<RangedEnemyBeliefSet>(moveTactic, MovePredicate));
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:88]:
```suggestion
            PrimitiveGoalStructure<RangedEnemyBeliefSet> targetReached = new(goal: new Goal<RangedEnemyBeliefSet>(moveTactic, MovePredicate));
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 35]
```diff
@@ -0,0 +1,139 @@
...
+
+            // Goal: Wait until killed by the enemy.
+            PrimitiveGoalStructure<RangedEnemyBeliefSet> waitGoal = new(goal: new Goal<RangedEnemyBeliefSet>(waitTactic, PlayerRespawnedPredicate));
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:91]:
```suggestion
            PrimitiveGoalStructure<RangedEnemyBeliefSet> killedByEnemy = new(goal: new Goal<RangedEnemyBeliefSet>(waitTactic, PlayerRespawnedPredicate));
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 36]
```diff
@@ -0,0 +1,139 @@
...
+
+            // Goal: Wait and check if enemy is not following player.
+            PrimitiveGoalStructure<RangedEnemyBeliefSet> waitGoal2 = new(goal: new Goal<RangedEnemyBeliefSet>(waitTactic, EnemyNotMovingPredicate));
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:94]:
```suggestion
            PrimitiveGoalStructure<RangedEnemyBeliefSet> notFollowedByEnemy = new(goal: new Goal<RangedEnemyBeliefSet>(waitTactic, EnemyNotMovingPredicate));
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 37]
```diff
@@ -0,0 +1,139 @@
...
+
+            AplibRunner testRunner = new(agent);
+
+            // Use the Assert class to test conditions.
+            // Use yield to skip a frame.
+            yield return testRunner.Test();
+
+            // Assert that the player has reached the target position.
+            Assert.IsTrue(condition: agent.Status == CompletionStatus.Success);
+            yield break;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:112]:
```suggestion
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs [hunk 38]
```diff
@@ -0,0 +1,139 @@
...
+            BdiAgent<RangedEnemyBeliefSet> agent = new(beliefSet, desire);
+
+            AplibRunner testRunner = new(agent);
+
+            // Use the Assert class to test conditions.
+            // Use yield to skip a frame.
+            yield return testRunner.Test();
+
+            // Assert that the player has reached the target position.
+            Assert.IsTrue(condition: agent.Status == CompletionStatus.Success);
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:111]: Nit
```suggestion
            Assert.AreEqual(agent.Status, CompletionStatus.Success);
```

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs:1]: I left some comments on the descriptiveness of variable names.

**Review by JensSteenmetz — DISMISSED** [review]
Overall looks good. Left some remarks, mostly on descriptiveness of variable names in the aplib test.

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs (hunk 9 above)
Comment (QuakeEye) [reply aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs:29]: The `EntitySound` script itself will not become obsolete, but using it in code will, as its functions will be called through the animation system.
This is why I recommend just leaving the usage of `EntitySound` as it is on main right now, to avoid merge conflicts, as its current usage is removed later on.
Furthermore, changing the usage of it is not part of your task :)

File: aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs (hunk 24 above)
Comment (QuakeEye) [reply aplib.net-demo/Assets/Testing/AplibTests/RangedEnemyTest.cs]: That might be the case, but what if we want to change this value? We would have to search through the code

**Review by SSelinn — DISMISSED** [review]

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs [hunk 39]
```diff
@@ -59,7 +70,7 @@ public override void UseWeapon()
             {
                 foreach (Collider collider in _targets)
                 {
-                    // Check if the collider with enemy tag has a Health component. If so, deal damage to it.
+                    // Check if the collider with enemy tag has a Health component. If so, deal damage to it.
```
Comment (QuakeEye) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs:73]: > added space

what

**Review by QuakeEye — CHANGES_REQUESTED** [review]

**Review by QuakeEye — APPROVED** [review]

**Review by SSelinn — APPROVED** [review]
